src/cls/queue/cls_rgw_queue.cc

```diff
@@ -0,0 +1,664 @@
+  try {
+    decode(urgent_data, bl_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_queue_urgent_data_ret(): failed to decode urgent data\n");
+    return -EINVAL;
+  }
+
+  bool is_last_entry = false;
+  in->clear();
+  out->clear();
+  ret = cls_queue_get_last_entry(hctx, in, out);
+  if (ret < 0) {
+    return ret;
+  }
+
+  cls_rgw_gc_obj_info info;
+  auto iter = out->cbegin();
+  try {
+    decode(info, iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_gc_queue_update_entry(): failed to decode entry\n");
+    return -EINVAL;
+  }
+
+  CLS_LOG(1, "INFO: tag of gc info is %s\n", info.tag.c_str());
+  if (info.tag == op.info.tag) {
+    is_last_entry = true;
+  }
+
+  //has_urgent_data signifies whether urgent data in queue has changed
+  bool has_urgent_data = false, tag_found = false;
+  //search in unordered map in head
+  auto it = urgent_data.urgent_data_map.find(op.info.tag);
+  if (it != urgent_data.urgent_data_map.end()) {
+    it->second = op.info.time;
+    tag_found = true;
+    has_urgent_data = true;
+  } else { //search in xattrs
+    bufferlist bl_xattrs;
+    int ret = cls_cxx_getxattr(hctx, "cls_queue_urgent_data", &bl_xattrs);
+    if (ret < 0 && (ret != -ENOENT && ret != -ENODATA)) {
+      CLS_LOG(0, "ERROR: %s(): cls_cxx_getxattrs() returned %d", __func__, ret);
+      return ret;
+    }
+    if (ret != -ENOENT && ret != -ENODATA) {
+      std::unordered_map<string,ceph::real_time> xattr_urgent_data_map;
+      auto iter = bl_xattrs.cbegin();
+      try {
+        decode(xattr_urgent_data_map, iter);
+      } catch (buffer::error& err) {
+        CLS_LOG(1, "ERROR: cls_gc_queue_remove(): failed to decode xattrs urgent data map\n");
+        return -EINVAL;
+      } //end - catch
+      if (xattr_urgent_data_map.size() > 0) {
+        auto found = xattr_urgent_data_map.find(info.tag);
+        if (found != xattr_urgent_data_map.end()) {
+          it->second = op.info.time;
+          tag_found = true;
+          //write the updated map back
+          bufferlist bl_map;
+          encode(xattr_urgent_data_map, bl_map);
+          ret = cls_cxx_setxattr(hctx, "cls_queue_urgent_data", &bl_map);
+          CLS_LOG(20, "%s(): setting attr: %s", __func__, "cls_queue_urgent_data");
+          if (ret < 0) {
+            CLS_LOG(0, "ERROR: %s(): cls_cxx_setxattr (attr=%s) returned %d", __func__, "cls_queue_urgent_data", ret);
+            return ret;
+          }
+        }
+      } // end - if xattrs size ...
+    }// end ret != ENOENT ...
+  }
+
+  bufferlist bl_urgent_data;
+  if (! tag_found) {
+    //try inserting in queue head
+    urgent_data.urgent_data_map.insert({op.info.tag, op.info.time});
+    urgent_data.num_head_urgent_entries += 1;
+    has_urgent_data = true;
+
+    //check if urgent data can fit in head
+    out->clear();
+    bool can_fit = false;
+    encode(urgent_data, bl_urgent_data);
+    ret = cls_queue_can_urgent_data_fit(hctx, &bl_urgent_data, out);
```

Comment:
rather than doing a test-encode to see whether it fits, can we just assign it to head.urgent_data and call a write_head() helper function? then have write_head() itself, after encoding, return -ENOSPC if it overflows head_size

Reply:
In this method, later cls_enqueue or cls_queue_update_last_entry is invoked, which also modifies the head. I observed that if we have two different calls to write the head from the same method, the second one overwrites the first, to work around this, I had clubbed urgent data with enqueue or update last entry.

Reply:
I think I had sent this in an email, but cant find it now :(

Reply:
okay, i see. on queue creation, maybe clients should specify the `max_urgent_data_size` explicitly, rather than the `head_size` which contains urgent data, cls_queue_head, etc. then cls_queue_can_urgent_data_fit() could just be a simple comparison of `bl_urgent_data.length() <= head.max_urgent_data_size` and wouldn't require serializing the head

i also think the refactoring proposed by the 'minimal low-level library' could help here, because the head write wouldn't be hidden inside of cls_enqueue(). instead, we would read the head ourselves at the beginning, pass a mutable reference to cls_queue_head to the internal queue_enqueue() function, then write the modified head back at the end. if that final call to queue_header_write() fails with ENOSPC because urgent_data is too big, we can spill over to xattrs and rewrite the head with its original urgent_data

Reply:
Will this write_head method be invoked by cls_rgw_queue.cc (or whatever its name will be finally), which will call the low-level library method cls_enqueue in cls_queue_src.cc? In that case should we expose the head of queue to the application (making use of queue)? I thought the internal book keeping that we do for a queue shouldn't be exposed to the application. What if they accidentally modify some values that they should not

Reply:
Instead of taking max_urgent_size as input from the client, cant we calculate that and store it in head based on the size of other entries and max queue size. 
Matt had suggested that a config variable be exposed which can be used to arrive at an optimum correct head size needed for gc while testing (rgw_gc_queue_head_size)

Reply:
> Instead of taking max_urgent_size as input from the client, cant we calculate that and store it in head based on the size of other entries and max queue size.

yep, that sounds great